### ISSUE TITLE
MacOSX ARM terminal size always is the default

### ISSUE THREAD

**Issue by apatrida** [ajalt/mordant#238]
Mordant 3.0.0

In default Terminal, ITerm2, and Intellij terminal on my ARM Mac, I always end up with terminal size of 79x24.

This program shows a bit more:
```
fun main() {
    val terminal = Terminal()
    terminal.updateSize()
    terminal.println(terminal.terminalInfo)
    terminal.println(terminal.terminalInterface.javaClass.name)
    terminal.println(terminal.size)
    println(TerminalInterfaceFfmMacos().getTerminalSize())
}
```

terminal size is 79x24 default, and the FFM interface returns `null`.  

```
TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=false, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=true)
com.github.ajalt.mordant.terminal.terminalinterface.ffm.TerminalInterfaceFfmMacos
79×24
null
```

In FFM `getTerminalSize` the code returns 0,0 for `val size = MacosCLibrary.winsize(arena)` and the `ioctl` check is false.

The JNA version also returns `null` for `getTerminalSize()`

```
TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=false, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=true)
com.github.ajalt.mordant.terminal.terminalinterface.jna.TerminalInterfaceJnaMacos
79×24
null
```

This is intended to work, no?

**Comment by ajalt** [comment]
Yes, that's supposed to work. Maybe the ioct is wrong on ARM? I don't have a mac to test on, but if you have a C compiler,  can you try this C code and see what it prints?

```c
#include <stdio.h>
#include <sys/ioctl.h>

int main() {
  struct winsize sz;
  printf("TIOCGWINSZ: 0x%lx\n", TIOCGWINSZ);
  printf("TIOCSWINSZ: 0x%lx\n", TIOCSWINSZ);
  ioctl(0, TIOCGWINSZ, &sz);
  printf("width: %i  height: %i\n", sz.ws_col, sz.ws_row);
  return 0;
}
```

**Comment by apatrida** [comment]
```
❯ /usr/bin/env stty size
39 114
```

I must have bonked the JNA test prior, but FFM definitely does not work.

JNA working: (Iterm, Terminal, not Intellij)

```
TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=true, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=false)
com.github.ajalt.mordant.terminal.terminalinterface.jna.TerminalInterfaceJnaMacos
114×39
114×39
foo
bar
```

while FFM continues to be wrong:

```
TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=true, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=false)
com.github.ajalt.mordant.terminal.terminalinterface.ffm.TerminalInterfaceFfmMacos
79×24
79×24
```

Output of C program in Iterm 2: (same terminal as above FFM test)

```
TIOCGWINSZ: 0x40087468
TIOCSWINSZ: 0x80087467
width: 114  height: 39
```

Terminal:
```
TIOCGWINSZ: 0x40087468
TIOCSWINSZ: 0x80087467
width: 230  height: 48
```

Intellij terminal:
```
TIOCGWINSZ: 0x40087468
TIOCSWINSZ: 0x80087467
width: 0  height: 0
```

Warp:
```
TIOCGWINSZ: 0x40087468
TIOCSWINSZ: 0x80087467
width: 153  height: 45
```

same terminal back to back:

```
Kotlin program, JDK 22, Gradle application plugin, --enable-native-access=ALL-UNNAMED

TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=true, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=false)
com.github.ajalt.mordant.terminal.terminalinterface.ffm.TerminalInterfaceFfmMacos
79×24

C program

TIOCGWINSZ: 0x40087468
TIOCSWINSZ: 0x80087467
width: 114  height: 39
```

**Comment by apatrida** [comment]
Ok, it is the application generated shell script that does something to kill it.  If I manually run the test class via the command line it works fine.

use the installed application script and it blocks the terminal, checking the parameters used.

**Comment by apatrida** [comment]
Ok, I thought it was the gradle generated build script, but it's not.  Running the resulting class by hand with the classpath has the same affect of being the default size.

I removed all other command line arguments to the JVM.  I don't think another dependency could affect this since you are calling directly via FFM.

**Comment by apatrida** [comment]
Created a test project with only Mordant, so it is not the dependencies.

**Comment by ajalt** [comment]
Thanks for the investigation! So the ioctl you shared is indeed different than the one being used in the FFM module; I updated the constant in #239, so once if you can, give that a try and if it still doesn't work we can reopen this issue.

**Comment by apatrida** [comment]
crashes. 

[hs_err_pid49592.log](https://github.com/user-attachments/files/17537574/hs_err_pid49592.log)

**Comment by apatrida** [comment]
trying this in C:

```
#include <stddef.h>
#include <stdio.h>
#include <sys/ioctl.h>

int main() {
    struct winsize sz;
    printf("TIOCGWINSZ: 0x%lx\n", TIOCGWINSZ);
    printf("sizeof(winsize): %lu\n", sizeof(struct winsize));
    // Print field offsets
    printf("offset ws_row: %lu\n", offsetof(struct winsize, ws_row));
    printf("offset ws_col: %lu\n", offsetof(struct winsize, ws_col));
    printf("offset ws_xpixel: %lu\n", offsetof(struct winsize, ws_xpixel));
    printf("offset ws_ypixel: %lu\n", offsetof(struct winsize, ws_ypixel));

    int ret = ioctl(0, TIOCGWINSZ, &sz);
    printf("ioctl return: %d\n", ret);
    printf("width: %i  height: %i\n", sz.ws_col, sz.ws_row);
    return 0;
}
```
results in:

```
TIOCGWINSZ: 0x40087468
sizeof(winsize): 8
offset ws_row: 0
offset ws_col: 2
offset ws_xpixel: 4
offset ws_ypixel: 6
ioctl return: 0
width: 0  height: 0
```

seems to match, no?

**Comment by apatrida** [comment]
Also `0x00005413L` for TIOCGWINSZ doesn't match my header files.  Unless mine are ARM only, then this value is wrong for MAC Intel as well.  I can dig up an Intel MAC to check this too.

**Comment by apatrida** [comment]
note, using JNA to do the same call results in the crash as well.  Is it possible the wrong dynamic library or something is being linked?

**Comment by apatrida** [comment]
from JLine:
https://github.com/jline/jline3/blob/master/terminal-ffm/src/main/java/org/jline/terminal/impl/ffm/CLibrary.java#L1056

but I'm not having luck on getting values back from it.  no crash, but 0,0 size. checking more tomorrow.

**Comment by apatrida** [comment]
Ok, found the problem.  whew!

This function
```
MethodHandlesHolder::handle
```

needs to add `Linker.Option.firstVariadicArg(2)` to the `linker.downcallHandle()` call as a linker option just for `ioctl` call.

effectively:

```
if (name == "ioctl") {
    linker.downcallHandle(
        it,
        FunctionDescriptor.of(resLayout, *argLayouts),
        Linker.Option.firstVariadicArg(2)
    )
} else {
    linker.downcallHandle(
        it,
        FunctionDescriptor.of(resLayout, *argLayouts),
    )
} 
```

but probably better to pass it in for some other way.

**Comment by apatrida** [comment]
and for MAC it should always be `private val TIOCGWINSZ = 0x40087468L` no matter the architecture.

**Comment by apatrida** [comment]
note both of these match JLine3:

https://github.com/jline/jline3/blob/master/terminal-ffm/src/main/java/org/jline/terminal/impl/ffm/CLibrary.java#L455

https://github.com/jline/jline3/blob/master/terminal-ffm/src/main/java/org/jline/terminal/impl/ffm/CLibrary.java#L1056

if that matters

**Comment by apatrida** [comment]
result:

```
❯ ./runsample progress
TerminalInfo(ansiLevel=TRUECOLOR, ansiHyperLinks=true, outputInteractive=true, inputInteractive=true, supportsAnsiCursor=false)
com.github.ajalt.mordant.terminal.terminalinterface.ffm.TerminalInterfaceFfmMacos
114×39 <---- YAY!
```

**Comment by apatrida** [comment]
@ajalt you can re-open?

**Comment by ajalt** [comment]
@apatrida Great find! I implemented your suggestion in #240.

**Comment by apatrida** [comment]
tested, works fine here.
